src/deploy/functions/cache/hash.ts

```diff
@@ -0,0 +1,55 @@
+import * as fs from "fs";
+import * as crypto from "crypto";
+import * as secrets from "../../../functions/secrets";
+import { Backend, Endpoint, EnvironmentVariables, allEndpoints } from "../backend";
+
+/**
+ * Retrieves the unique hash given a {@link Backend} and sourceFile.
+ * @param backend Backend of a set of functions
+ * @param sourceFile Packaged file contents of functions
+ */
+export async function getBackendHash(backend: Backend, sourceFile?: string) {
+  const hash = crypto.createHash("sha256");
+
+  // If present, hash the contents of the source file
+  if (sourceFile) {
+    await new Promise((resolve, reject) => {
+      const readStream = fs.createReadStream(sourceFile);
+      readStream.on("error", (err) => reject(err));
+      readStream.on("data", (chunk) => chunk && hash.push(chunk));
+      readStream.on("end", () => resolve(hash.digest("hex")));
+    });
+  }
+
+  // Hash the contents of the dotenv variables
+  if (hasEnvironmentVariables(backend.environmentVariables)) {
+    hash.push(JSON.stringify(backend.environmentVariables));
+  }
+
+  // Hash the secret versions.
+  const secretVersions = getSecretVersions(backend);
+  if (hasSecretVersions(secretVersions)) {
+    hash.push(JSON.stringify(secretVersions));
+  }
+
+  return hash.read().toString("hex");
+}
+
+function hasEnvironmentVariables(environmentVariables: EnvironmentVariables) {
+  return Object.keys(environmentVariables).length;
+}
+
+function hasSecretVersions(secretVersions: Record<string, string>) {
+  return Object.keys(secretVersions).length;
+}
+
+// Hash the secret versions.
+function getSecretVersions(backend: Backend): Record<string, string> {
+  const endpointsList: Endpoint[] = allEndpoints(backend);
+  return secrets.of(endpointsList).reduce((memo, { secret, version }) => {
+    if (version) {
+      memo[secret] = version;
```

Comment:
Is there a bug here where if I'm using different versions of secret (for some reason) for different functions, I loose info?

Reply:
This is totally a bug. Good thing this function is being blown away by an endpoint version. 👍